### PR TITLE
Add js2-jsx-mode

### LINKED ISSUE #140: Highlighting & indenting embedded XML (E4X)?
Hello,

It seems that js2-mode already has a reasonable parser of embedded XML in JavaScript - but I can't see if there's highlighting & indenting.

To be more precise, AFAIK `js2-record-face` function should be called for each highlighted token, and I can't see if it's called for XML tokens too (`js2-XML` and `js2-XMLATTR`) - so did I understand correctly that there's no highlighting for E4X yet?

If so, I'd try to add it - IMO it would be a very useful feature. I know that E4X was deprecated, but embedded XML is heavily used by [React](http://facebook.github.io/react/) in form of JSX, and the syntax essentially is the same as E4X, so no changes would be required in the parser.

Or, perhaps, that's a problem on my side, and `js2-mode` already works great with React? :)

Please let me know.
Thanks!

### PR DESCRIPTION
**Pull request by jacksonrayhamilton**
Complements the [latest change to master](http://git.savannah.gnu.org/cgit/emacs.git/commit/?id=958da7ff63d1d647f45fd649654136da78529324). Partially resolves #140.

### CODE REVIEW

File: js2-mode.el [hunk 1]
```diff
@@ -11373,6 +11373,17 @@ Selecting an error will jump it to the corresponding source-buffer error.
...
 
+;; We may eventually want js2-jsx-mode to derive from js-jsx-mode, but that'd be
+;; a bit more complicated and it doesn't net us much yet.
+;;;###autoload
+(define-derived-mode js2-jsx-mode js2-mode "JSX-IDE"
+  "Major mode for editing JSX code.
+Like `js-jsx-mode', which see.
+Indentation support requires Emacs 25+ (or the latest js.el)."
+  ;; Feature-detect JSX indentation support (in js-mode from Emacs 25+).
+  (if (fboundp 'js-jsx-indent-line)
```
Comment (dgutov) [inline js2-mode.el]: What if the user's Emacs is too old? Will it silently do nothing?

Seems like a recipe for confusion.

Comment (jacksonrayhamilton) [inline js2-mode.el]: We could print the message `Indentation support requires Emacs 25+ (or the latest js.el)`?

Or copy over the new indentation code to js2-old-indent.el?

Comment (dgutov) [inline js2-mode.el]: > Or copy over the new indentation code to js2-old-indent.el?

That's an option.

Or we just ask people to use `js2-minor-mode` with `js-jsx-mode`.

Or we put this mode in a separate file, and signal an error when it's loaded when the version of Emacs isn't high enough. Or simply signal this error from inside of `js2-jsx-mode`, and keep it in this file.

I think an error is better than a warning message.

Comment (jacksonrayhamilton) [inline js2-mode.el]: Decided to copy it over, as that should be easiest for users.

Bu

**Comment by dgutov** [comment]
All right. Please add a NEWS entry as well.

**Comment by dgutov** [comment]
Should we also have instructions how to use it (adding it to `auto-mode-alist`), in the top commentary?

**Comment by dgutov** [comment]
Fantastic, thanks!

**Comment by istinspring** [comment]
so now it's possible to use js2-mode for ReactJS?

**Pull request merged**